### PR TITLE
feat: add githubgen tool

### PR DESCRIPTION
**Pull request by mowies**
Fixes https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/36691

### CODE REVIEW

File: githubgen/codeowners.go [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+	"sort"
+	"strings"
+
+	"github.com/google/go-github/v66/github"
+)
+
+const allowlistHeader = `# Code generated by githubgen. DO NOT EDIT.
+#####################################################
+#
+# List of components in OpenTelemetry Collector Contrib
```
Comment (atoulme) [inline githubgen/codeowners.go:19]: hardcoded to contrib

File: githubgen/codeowners.go [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+const unmaintainedHeader = `
+
+## UNMAINTAINED components
+
+`
+
+const codeownersHeader = `# Code generated by githubgen. DO NOT EDIT.
+#####################################################
+#
+# List of codeowners for OpenTelemetry Collector Contrib
```
Comment (atoulme) [inline githubgen/codeowners.go:50]: hardcoded, we might need to make that configurable.

File: githubgen/codeowners.go [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+#
+# Learn about membership in OpenTelemetry community:
+# https://github.com/open-telemetry/community/blob/main/guides/contributor/membership.md
+#
+#
+# Learn about CODEOWNERS file format:
+# https://help.github.com/en/articles/about-code-owners
+#
+
+* @open-telemetry/collector-contrib-approvers
```
Comment (atoulme) [inline githubgen/codeowners.go:62]: same

File: githubgen/codeowners.go [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+`
+
+const distributionCodeownersHeader = `
+#####################################################
+#
+# List of distribution maintainers for OpenTelemetry Collector Contrib
```
Comment (atoulme) [inline githubgen/codeowners.go:68]: also hardcoded

File: githubgen/codeowners.go [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+
+	unmaintainedCodeowners := unmaintainedHeader
+	currentFirstSegment := ""
+LOOP:
+	for _, key := range data.folders {
+		m := data.components[key]
+		for stability := range m.Status.Stability {
+			if stability == unmaintainedStatus {
+				unmaintainedList += key + "/\n"
+				unmaintainedCodeowners += fmt.Sprintf("%s/%s @open-telemetry/collector-contrib-approvers \n", key, strings.Repeat(" ", data.maxLength-len(key)))
```
Comment (atoulme) [inline githubgen/codeowners.go:142]: hardcoded as well here.

File: githubgen/codeowners.go [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+			if firstSegment != currentFirstSegment {
+				currentFirstSegment = firstSegment
+				codeowners += "\n"
+			}
+			owners := ""
+			for _, owner := range m.Status.Codeowners.Active {
+				owners += " "
+				owners += "@" + owner
+			}
+			codeowners += fmt.Sprintf("%s/%s @open-telemetry/collector-contrib-approvers%s\n", key, strings.Repeat(" ", data.maxLength-len(key)), owners)
```
Comment (atoulme) [inline githubgen/codeowners.go:162]: hardcoded as well here

**Comment by mowies** [comment]
@atoulme thanks for the review. I will add at least a sanity check unit test in this PR which will already need some slight refactoring and then do any further changes in follow-ups.

**Comment by mowies** [comment]
@atoulme this is basically ready but i can't get `moq` to work on windows for some reason 🤦🏼

**Comment by mx-psi** [comment]
@pjanotti If you have the time, could you take a quick look at the failing CI check and try to help us figure out why it is failing?

**Comment by pjanotti** [comment]
@mowies the issue can be solved by creating a `moq.exe` at the `.tools` directory. When in the make shell env the `bash.exe` takes care of that, but, I suspect that `go generate` (as it should) uses the Windows API. A quick workaround will be to `cp .\.tools\moq .\.tools\moq.exe`

**Comment by mowies** [comment]
@pjanotti wow thanks for the help, I don't think I would have found that by myself at all 😅 
seems like everything works now 🙌🏼

File: githubgen/issuetemplates.go [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+
+func folderToShortName(folder string) string {
+	if folder == "internal/coreinternal" {
+		return "internal/core"
+	}
+	path := strings.Split(folder, "/")
+	switch path[0] {
+	case "receiver", "exporter", "extension", "processor", "connector":
+		path[1] = strings.TrimSuffix(strings.TrimSuffix(strings.TrimSuffix(strings.TrimSuffix(strings.TrimSuffix(strings.TrimSuffix(path[1], "internal"), "extension"), "exporter"), "connector"), "processor"), "receiver")
+		path[len(path)-1] = strings.TrimSuffix(strings.TrimSuffix(strings.TrimSuffix(strings.TrimSuffix(strings.TrimSuffix(strings.TrimSuffix(path[len(path)-1], "internal"), "extension"), "exporter"), "connector"), "processor"), "receiver")
```
Comment (dmathieu) [inline githubgen/issuetemplates.go:29]: This looks like it shouldn't be hardcoded, but configurable per repository?

**Comment by mowies** [comment]
@atoulme @dmathieu thanks for the reviews! I will implement the feedback along with more unit tests in a follow up PR so that all that gets easier to review.

**Review by mx-psi — APPROVED** [review]
Approving based on the comment above

**Comment by mowies** [comment]
follow up PR that addresses all the feedback here and more is ready for review: https://github.com/open-telemetry/opentelemetry-go-build-tools/pull/655

**Pull request merged**